File: backend/services/stigmer-server/pkg/domain/mcpserver/oauth/refresh.go
package oauth

import (
	"context"
	"fmt"
	"time"

	"github.com/rs/zerolog/log"
)

// RefreshResult holds the outcome of a token refresh attempt.
type RefreshResult struct {
	Refreshed       bool
	NewAccessToken  string
	NewRefreshToken string
	NewExpiresAt    int64
}

// RefreshTokenIfExpired checks if the access token for a given OAuthGrant
// is expired, and if so, uses the refresh token to obtain a new one.
//
// Returns a RefreshResult indicating whether a refresh occurred and the
// new token values. The caller is responsible for updating the personal
// environment and the OAuthGrant record with the new values.
//
// Parameters:
//   - grant: the OAuthGrant record (must not be nil)
//   - currentRefreshToken: the decrypted refresh token from the personal env
//   - clientSecret: decrypted client_secret (empty for DCR/public clients)
//
// Returns:
//   - RefreshResult with new token values if refreshed
//   - nil result if the token is not expired
//   - error if the refresh failed (caller should return re-auth error to user)
func RefreshTokenIfExpired(
	ctx context.Context,
	grant *OAuthGrant,
	currentRefreshToken string,
	clientSecret string,
) (*RefreshResult, error) {
	if grant.AccessTokenExpiresAt == 0 {
		// Token does not expire (e.g., long-lived tokens from Notion/Slack)
		return &RefreshResult{Refreshed: false}, nil
	}

	now := time.Now().Unix()
	// Add a 60-second buffer to refresh slightly before expiry
	if now < grant.AccessTokenExpiresAt-60 {
		return &RefreshResult{Refreshed: false}, nil
	}

	if currentRefreshToken == "" {
		return nil, fmt.Errorf(
			"access token for resource '%s' has expired and no refresh token is available. "+
				"Please re-authenticate via OAuth Connect",
			grant.ResourceID,
		)
	}

	log.Info().
		Str("resource_id", grant.ResourceID).
		Str("resource_kind", grant.ResourceKind).
		Int64("expired_at", grant.AccessTokenExpiresAt).
		Str("token_endpoint", grant.TokenEndpoint).
		Msg("Access token expired, refreshing via refresh_token grant")

	tokenResp, err := RefreshToken(
		ctx,
		grant.TokenEndpoint,
		currentRefreshToken,
		grant.ClientID,
		clientSecret,
	)
	if err != nil {
		return nil, fmt.Errorf(
			"token refresh failed for resource '%s': %w. "+
				"Please re-authenticate via OAuth Connect",
			grant.ResourceID, err,
		)
	}

	var newExpiresAt int64
	if tokenResp.ExpiresIn > 0 {
		newExpiresAt = time.Now().Unix() + tokenResp.ExpiresIn
	}

	newRefreshToken := currentRefreshToken
	if tokenResp.RefreshToken != "" {
		newRefreshToken = tokenResp.RefreshToken
	}

	log.Info().
		Str("resource_id", grant.ResourceID).
		Str("resource_kind", grant.ResourceKind).
		Int64("new_expires_at", newExpiresAt).
		Bool("refresh_token_rotated", tokenResp.RefreshToken != "").
		Msg("Token refresh successful")

	return &RefreshResult{
		Refreshed:       true,
		NewAccessToken:  tokenResp.AccessToken,
		NewRefreshToken: newRefreshToken,
		NewExpiresAt:    newExpiresAt,
	}, nil
}

File: backend/services/stigmer-server-ts/src/domain/mcpserver/oauth/refresh.ts
/**
 * Pre-flight token refresh — ports pkg/domain/mcpserver/oauth/refresh.go.
 * Consumed by the agentexecution EC builder's OAuth injection (#17); the
 * connect-lane refresh callers arrive with #19.
 */
import type { OAuthGrant } from "../../../store/interface.js";
import type { Logger } from "../../../boot/logger.js";
import { refreshToken } from "./token.js";

/** The outcome of a refresh attempt (Go RefreshResult). */
export interface RefreshResult {
  refreshed: boolean;
  newAccessToken: string;
  newRefreshToken: string;
  /** Unix seconds; 0 = does not expire. */
  newExpiresAt: number;
}

/**
 * Refresh slightly before expiry — Go's 60-second buffer, named per the
 * guidelines' semantic-constant rule.
 */
export const REFRESH_EXPIRY_BUFFER_SECONDS = 60;

/**
 * Checks whether the grant's access token is expired and, if so, uses the
 * refresh token to obtain a new one. The caller updates the managed
 * environment and the grant record with the returned values.
 *
 * clientSecret is empty for DCR/public clients; tokenAuthMethod selects
 * how a non-empty secret is presented (empty falls back to Basic).
 * Returns refreshed=false when the token has no expiry (long-lived
 * Notion/Slack-style tokens) or is not yet within the buffer; throws when
 * the refresh is needed but impossible or fails (the caller surfaces a
 * re-auth error).
 */
export async function refreshTokenIfExpired(
  grant: OAuthGrant,
  currentRefreshToken: string,
  clientSecret: string,
  tokenAuthMethod: string,
  logger: Logger,
  fetchImpl: typeof fetch = fetch,
): Promise<RefreshResult> {
  if (grant.accessTokenExpiresAt === 0) {
    // Token does not expire.
    return notRefreshed();
  }

  const now = Math.floor(Date.now() / 1000);
  if (now < grant.accessTokenExpiresAt - REFRESH_EXPIRY_BUFFER_SECONDS) {
    return notRefreshed();
  }

  if (currentRefreshToken === "") {
    throw new Error(
      `access token for resource '${grant.resourceId}' has expired and no refresh token is available. ` +
        "Please re-authenticate via OAuth Connect",
    );
  }

  logger.info("Access token expired, refreshing via refresh_token grant", {
    resourceId: grant.resourceId,
    resourceKind: grant.resourceKind,
    expiredAt: grant.accessTokenExpiresAt,
    tokenEndpoint: grant.tokenEndpoint,
  });

  let tokenResponse;
  try {
    tokenResponse = await refreshToken(
      grant.tokenEndpoint,
      currentRefreshToken,
      grant.clientId,
      clientSecret,
      tokenAuthMethod,
      fetchImpl,
    );
  } catch (error) {
    throw new Error(
      `token refresh failed for resource '${grant.resourceId}': ${error instanceof Error ? error.message : String(error)}. ` +
        "Please re-authenticate via OAuth Connect",
    );
  }

  let newExpiresAt = 0;
  if (tokenResponse.expiresIn > 0) {
    newExpiresAt = Math.floor(Date.now() / 1000) + tokenResponse.expiresIn;
  }

  const newRefreshToken =
    tokenResponse.refreshToken !== ""
      ? tokenResponse.refreshToken
      : currentRefreshToken;

  logger.info("Token refresh successful", {
    resourceId: grant.resourceId,
    resourceKind: grant.resourceKind,
    newExpiresAt,
    refreshTokenRotated: tokenResponse.refreshToken !== "",
  });

  return {
    refreshed: true,
    newAccessToken: tokenResponse.accessToken,
    newRefreshToken,
    newExpiresAt,
  };
}

function notRefreshed(): RefreshResult {
  return {
    refreshed: false,
    newAccessToken: "",
    newRefreshToken: "",
    newExpiresAt: 0,
  };
}
